nested_lookup/lookup_api.py

```diff
@@ -29,28 +31,134 @@ def _nested_delete(document, key):
     """
     if isinstance(document, list):
         for list_items in document:
-            _nested_update(document=list_items, key=key, value=value)
+            _nested_update(document=list_items, key=key, value=value, val_len = val_len, run = run)
     elif isinstance(document, dict):
         if document.get(key):
-            document[key] = value
+            # check if a value with the coresponding index exists and use it otherwise recycle the intially given value
+            if run < val_len:
+                val = value[run]  
+            else: 
+                run = 0
+                val = value[run] 
+            document[key] = val
+            run = run + 1
         for dict_key, dict_value in iteritems(document):
-            _nested_update(document=dict_value, key=key, value=value)
+            _nested_update(document=dict_value, key=key, value=value, val_len = val_len, run = run)
+    return document
+
+def nested_alter(document:dict, key:str, callback_function = None, function_parameters:list = None, in_place:bool = True):
+    """
+    Method to alter all values of the occurences of the key "key".
+    The provided callback_function is used to alter the scalar values
+    Args:
+        document: Might be List of Dicts (or) Dict of Lists (or)
+            Dict of List of Dicts etc...
+        key: Key to update the value
+        callback_function (Function): A callback function which alters a scalar value.
+            HINT: You should be aware that not every element might be of the same type, please check this in your function!
+        function_parameters (list):
+            If the callback_function has additional input arguments except the scalar value, please specify those in this list.
+        in_place (bool): 
+            True: modify the dict in place; 
+            False: create a deep copy of the dict and modify it
+            Defaults to False
+    Return:
+        (dict) Returns a document that has updated key, value pair.
+    """
+ # check if a list or scalar value is provided and create a list from the scalar value
+    # check the length of the list and provide it to _nested_update
+    if isinstance(key, list):
+        key_len = len(key)
+    else: 
+        key  = [key]
+        key_len = len(key)
+
+    if not in_place:
+        document = copy.deepcopy(document)
+    return _nested_alter(document=document, keys=key, callback_function=callback_function, function_parameters = function_parameters, in_place = in_place, key_len = key_len)
+
+def _nested_alter(document:dict, keys, callback_function, function_parameters:list, in_place:bool, key_len:int):
+    """
+    """
+    # return data if no callback_function is provided
+    if callback_function is None:
+        warnings.warn("Please provide a callback_function to nested_alter().")
+        return document
+
+    def _call_callback(value_list:list, callback_function, function_parameters):
```

Comment:
Can we can have this method `_call_callback ` outside the scope of method `_nested_alter`?

Reply:
Done